cowbird/services/service_factory.py

```diff
@@ -0,0 +1,59 @@
+from typing import TYPE_CHECKING
+
+import six
+
+from cowbird.config import get_all_configs
+from cowbird.utils import SingletonMeta, get_config_path
+
+if TYPE_CHECKING:
+    from typing import List
+
+    from cowbird.services.service import Service
+
+VALID_SERVICES = ["Catalog", "Geoserver", "Magpie", "Nginx", "Thredds",
+                  "FileSystem"]
+
+
+@six.add_metaclass(SingletonMeta)
+class ServiceFactory:
+    """
+    Create service instance using service name.
+    """
+
+    def __init__(self):
+        config_path = get_config_path()
+        self.services_cfg = get_all_configs(config_path, "services", allow_missing=True)[0]
+        self.services = {}
+
+    def get_service(self, name):
+        # type: (ServiceFactory, str) -> Service
+        """
+        Instantiates a `Service` implementation using its name if it doesn't exist or else returns the existing one from
+        cache.
+        """
+        try:
+            return self.services[name]
+        except KeyError:
+            # pylint: disable=W0641,unused-import
+            from cowbird.services.impl.catalog import Catalog  # noqa: F401
+            from cowbird.services.impl.filesystem import FileSystem  # noqa: F401
+            from cowbird.services.impl.geoserver import Geoserver  # noqa: F401
+            from cowbird.services.impl.magpie import Magpie  # noqa: F401
+            from cowbird.services.impl.nginx import Nginx  # noqa: F401
+            from cowbird.services.impl.thredds import Thredds  # noqa: F401
```

Comment:
I tried to remove those and I still get complains. Since the factory creates the Service from a string I understand that those imports are seen as unused.

Reply:
Is it possible that it is caused by the installed pylint version you run it with?
I find in PyCharm that types are checked independently. 
So if `Geoserver`, `Magpie`, etc.  are somewhere in the `# type: ...`, they are found, and otherwise they are greyed out.

Reply:
Example: 

![image](https://user-images.githubusercontent.com/19194484/116936740-3efe5980-ac36-11eb-8d95-aca56cee24c5.png)

Reply:
With the `noqa`, they are not flagged, so they can't be spotted at first glance. 
Also makes it troublesome to always add that `noqa` part.

Reply:
It's not a type thing, there is no reference at all to those implementations in this file. The factory get a reference to them at runtime and if they are not imported, it crashes. The only way I can think of would be to replace my typing with an explicit list of service like this : 
`# type: (ServiceFactory, str) -> Any[Catalog, FileSystem, Geoserver, Magpie, Nginx, Thredds]`
But then I need to import the modules properly (not under TYPE_CHECKING) and to avoid recursive imports I need to do the import in the function before the typing! Not pretty at all ;)

Reply:
I could get rid of import using a dynamic import like this if you prefer : 
```
module = importlib.import_module(".".join(["cowbird.services.impl", name.lower()]))
cls = getattr(module, name)
```

Reply:
Oh, I see. 
I'm so used to seeing `# noqa: F401`  right next to imports for typing that I didn't realize it was actual import tests/retry. 
Ignore my previous comments. Good enough if it works.

Reply:
I choose to switch to dynamic import. I find it cleaner.